File: Python/leetcodepractice/study_plan_leetcode75/LC61.py
# https://leetcode.com/problems/rotate-list/

from typing import Optional

from leetcodepractice.data_structure_elements import ListNode


class LC61:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        length = self._make_cycle_and_return_length(head)
        step = length - k % length
        new_tail = None
        for i in range(step):
            new_tail = head
            head = head.next
        new_tail.next = None
        return head

    @staticmethod
    def _make_cycle_and_return_length(head: Optional[ListNode]) -> int:
        length = 1
        tail = head
        while tail.next is not None:
            tail = tail.next
            length += 1
        tail.next = head
        return length

File: lib/src/main/java/leetcodepractice/LC61.java
package leetcodepractice;

import leetcodepractice.datastructure.ListNode;

// https://leetcode.com/problems/rotate-list/
// Advance version: https://www.algoexpert.io/questions/shift-linked-list
public class LC61 {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        int length = makeCycleAndReturnLength(head);
        int step = (length - (k % length)) % length;
        ListNode newTail = null;
        for (int i = 0; i < step; i++) {
            newTail = head;
            head = head.next;
        }
        newTail.next = null;
        return head;
    }

    private int makeCycleAndReturnLength(ListNode head) {
        int length = 1;
        ListNode tail = head;
        while (tail.next != null) {
            tail = tail.next;
            length++;
        }
        tail.next = head;
        return length;
    }
}
